### PR TITLE
Add groupMap and groupMapReduce extensions

### LINKED ISSUE #11: Add groupMap and groupMapReduce
So that users can migrate the following pattern:

~~~ scala
iterable.groupBy(f).mapValues(g).toMap
~~~

Into just:

~~~ scala
iterable.groupMap(f)(g)
~~~

### PR DESCRIPTION
**Pull request by NthPortal**
Fixes #11

### CODE REVIEW

**Comment by NthPortal** [comment]
I think there's some MiMa twiddling that needs to happen, but I believe this code should pretty much work

File: compat/src/main/scala-2.11_2.12/scala/collection/compat/PackageShared.scala [hunk 1]
```diff
@@ -243,6 +250,40 @@ class TraversableExtensionMethods[A](private val self: c.Traversable[A]) extends
   def iterableFactory: GenericCompanion[Traversable] = self.companion
 }
 
+class TraversableLikeExtensionMethods[A, Repr](private val self: c.GenTraversableLike[A, Repr])
+    extends AnyVal {
+
+  def groupMap[K, B, VThat, That](key: A => K)(f: A => B)(
+      implicit traversable: IsTraversableLikeAux[A, Repr],
+      innerBf: CanBuildFrom[Repr, B, VThat],
+      outerBf: CanBuildFrom[c.GenMap[K, Repr], (K, VThat), That]): That = {
```
Comment (julienrf) [inline compat/src/main/scala-2.11_2.12/scala/collection/compat/PackageShared.scala]: I think we should fix the return type to `immutable.Map[K, VThat]`, like we do in the scala-library.

File: compat/src/main/scala-2.11_2.12/scala/collection/compat/PackageShared.scala [hunk 2]
```diff
@@ -243,6 +250,40 @@ class TraversableExtensionMethods[A](private val self: c.Traversable[A]) extends
...
+      val innerB = innerBf(repr)
+      for (a <- traversable.conversion(repr)) innerB += f(a)
+      outerB.+=((k, innerB.result()))
+    }
+    outerB.result()
+  }
+
+  def groupMapReduce[K, B, That](key: A => K)(f: A => B)(reduce: (B, B) => B)(
+      implicit traversable: IsTraversableLikeAux[A, Repr],
+      bf: CanBuildFrom[c.GenMap[K, Repr], (K, B), That]): That = {
```
Comment (julienrf) [inline compat/src/main/scala-2.11_2.12/scala/collection/compat/PackageShared.scala]: Here too, we could simply fix the return type to `immutable.Map[K, B]`.

File: compat/src/main/scala-2.11_2.12/scala/collection/compat/PackageShared.scala (hunk 1 above)
Comment (NthPortal) [reply compat/src/main/scala-2.11_2.12/scala/collection/compat/PackageShared.scala]: I don't have a problem with that; however, it's not what `groupBy` does

Comment (julienrf) [reply compat/src/main/scala-2.11_2.12/scala/collection/compat/PackageShared.scala]: I was referring to [these implementations](https://github.com/scala/scala/blob/v2.13.1/src/library/scala/collection/Iterable.scala#L538-L630), I think that using as few `CanBuildFrom` as possible will make the code more readable :)

Comment (NthPortal) [reply compat/src/main/scala-2.11_2.12/scala/collection/compat/PackageShared.scala]: I rewrote it to just use the 2.13 implementation, which massively simplified things

File: compat/src/main/scala-2.11_2.12/scala/collection/compat/PackageShared.scala [hunk 3]
```diff
@@ -30,6 +30,8 @@ private[compat] trait PackageShared {
    */
   type Factory[-A, +C] = CanBuildFrom[Nothing, A, C]
 
+  type IsTraversableLikeAux[A1, Repr] = IsTraversableLike[Repr] { type A = A1 }
```
Comment (julienrf) [inline compat/src/main/scala-2.11_2.12/scala/collection/compat/PackageShared.scala]: You don’t seem to be using this type alias anymore.

**Comment by julienrf** [comment]
Unfortunately, this PR is not backward compatible (see https://stackoverflow.com/a/18369508/561721 for an explanation of why).

I think we could create a `2.x` branch where we could apply backward-compatible changes (if needed).

And branch `master` could be the place where we prepare the 3.x release…

What do others think?

**Comment by lrytz** [comment]
Maybe the `toTraversableLikeExtensionMethods` method can be duplicated and added directly to the `compat` objects, instead of the `PackageShared` super trait? That would avoid the binary incompatibility on 2.11.

**Comment by NthPortal** [comment]
for future versions, can `PackageShared` be an `abstract class` rather than a `trait` (thus sidestepping the problem)?

**Comment by NthPortal** [comment]
@lrytz yup, I'll do that

**Comment by julienrf** [comment]
@NthPortal Yes, that should work (see #292)

File: compat/src/main/scala-2.11_2.12/scala/collection/compat/PackageShared.scala (hunk 3 above)
Comment (NthPortal) [reply compat/src/main/scala-2.11_2.12/scala/collection/compat/PackageShared.scala]: good catch!

**Comment by NthPortal** [comment]
cleaned up, squashed, and ready for merge

**Pull request merged**